### PR TITLE
Fix CookieHeaderParserShared throws exception when last cookie contains invalid character #45014

### LINKED ISSUE #45014: CookieHeaderParserShared throws exception when last cookie contains invalid character
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Cookie String: keya=valuea; keyb=valueb; errorcookie=dd,:("sa;
the last part of cookie string above is an invalid cookie and the whole cookie string ends with `;`，this will throw exception like below, and all cookies fails to be read.

Exception: Nullable object must have a value.

   at Microsoft.Net.Http.Headers.CookieHeaderParserShared.TryParseValues(StringValues values, IDictionary`2 store, Boolean enableCookieNameEncoding, Boolean supportsMultipleValues)
   at Microsoft.AspNetCore.Http.RequestCookieCollection.ParseInternal(StringValues values, Boolean enableCookieNameEncoding)
   at Microsoft.AspNetCore.Http.Features.RequestCookiesFeature.get_Cookies()
... try get value from cookies

Cookie String: keya=valuea; keyb=valueb; errorcookie=dd,:("sa
the last part of this cookie string above is an invalid cookie and the whole cookie string ends without `;`，this will not throw exception, the parser will ignore `errorcookie`

### Expected Behavior

error cookie be ignored, and normal cookie still readable.

### Steps To Reproduce

_No response_

### Exceptions (if any)

Nullable object must have a value.

   at Microsoft.Net.Http.Headers.CookieHeaderParserShared.TryParseValues(StringValues values, IDictionary`2 store, Boolean enableCookieNameEncoding, Boolean supportsMultipleValues)
   at Microsoft.AspNetCore.Http.RequestCookieCollection.ParseInternal(StringValues values, Boolean enableCookieNameEncoding)
   at Microsoft.AspNetCore.Http.Features.RequestCookiesFeature.get_Cookies()

### .NET Version

7.0

### Anything else?

visual studio 2022 17.4.0

### PR DESCRIPTION
**Pull request by korteksz**
Remove supportsMultipleValues parameter from CookieHeaderParserShared.TryParseValues() method to prevent parsing strings wrong which contain separator characters. Reported by this issue: #45014

# Fix CookieHeaderParserShared throws exception when last cookie contains invalid character #45014

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description
Remove supportsMultipleValues parameter from CookieHeaderParserShared.TryParseValues() method to prevent parsing strings wrong (partly) which contain separator characters. 

More details about the problem: https://github.com/dotnet/aspnetcore/issues/45014#issuecomment-1312540487

Fixes #45014

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=45127) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=45127) before we can accept your contribution.<br/><br/>:x: korteksz [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=45127)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=45127) it.</sub>

**Comment by Tratcher** [comment]
> Remove supportsMultipleValues parameter from CookieHeaderParserShared.TryParseValues() method to prevent parsing strings wrong (partly) which contain separator characters.

I don't think this is the right aproach, it seems to have caused many cookie related failures in the SignalR tests.
https://dev.azure.com/dnceng-public/public/_build/results?buildId=85624&view=ms.vss-test-web.build-test-results-tab&runId=1791430&paneView=debug&resultId=100695

supportsMultipleValues means that a single header string might contain multiple cookies. That's fine, especially since it's common for cookie headers to be concatenated incorrectly.

I think there's a lower issue where something is returning true but null, where it should be returning false. Probably here:
https://github.com/dotnet/aspnetcore/blob/860b12b64d49aa7ffae534601844dd54d0363ebf/src/Http/Shared/CookieHeaderParserShared.cs#L29

**Comment by BrennanConroy** [comment]
Can you please review and sign the CLA? https://github.com/dotnet/aspnetcore/pull/45127#issuecomment-1317296994

**Comment by korteksz** [comment]
@BrennanConroy I have already signed the CLA from my mobile phone and when I try to do it from the desktop, then I got  `MongoTopologyClosedError: Topology is closed` error even if I click the recheck link below.

**Comment by korteksz** [comment]
> > Remove supportsMultipleValues parameter from CookieHeaderParserShared.TryParseValues() method to prevent parsing strings wrong (partly) which contain separator characters.
> 
> I don't think this is the right aproach, it seems to have caused many cookie related failures in the SignalR tests. https://dev.azure.com/dnceng-public/public/_build/results?buildId=85624&view=ms.vss-test-web.build-test-results-tab&runId=1791430&paneView=debug&resultId=100695
> 
> supportsMultipleValues means that a single header string might contain multiple cookies. That's fine, especially since it's common for cookie headers to be concatenated incorrectly.
> 
> I think there's a lower issue where something is returning true but null, where it should be returning false. Probably here:
> 
> https://github.com/dotnet/aspnetcore/blob/860b12b64d49aa7ffae534601844dd54d0363ebf/src/Http/Shared/CookieHeaderParserShared.cs#L29

Thanks for pointing this out. I assumed that this won't be the right solution. That changeset is revoked now and I've got a different one commited and the pipelines are green as well. 

The problem is that if you have a cookie like errorcookie=dd,:("sa; then it gets parsed currently like:
Key: errorcookie Value: dd -> because it parses the cookie until it doesn't find a non-cookie character, in this case the ',' char. Then the while loop keeps iterating further until the last character which is a ';'. As far as this is a separator and we are at the end of the input value and because supportsMultipleValues is true we return true at the end, however our parsedName and parsedValue is null. To avoid that I added a check to ignore separators at the end of the cookie value. Hope this one is better than the previous :)

File: src/Http/Shared/CookieHeaderParserShared.cs [hunk 1]
```diff
@@ -107,6 +107,13 @@ private static int GetNextNonEmptyOrWhitespaceIndex(StringSegment input, int sta
         // If we have a separator, skip the separator and all following whitespaces. If we support
         // empty values, continue until the current character is neither a separator nor a whitespace.
         separatorFound = true;
+
+        // We accept single separator, but separator at the end of a value is ignored.
+        if (input.Length > 1 && current + 1 == input.Length)
```
Comment (Tratcher) [inline src/Http/Shared/CookieHeaderParserShared.cs:112]: I tried adding these tests and hit the original exception:
```
    [InlineData(",", null)]
    [InlineData(";", null)]
```

File: src/Http/Shared/CookieHeaderParserShared.cs [hunk 2]
```diff
@@ -28,8 +28,15 @@ public static bool TryParseValues(StringValues values, IDictionary<string, strin
             {
                 if (TryParseValue(value, ref index, supportsMultipleValues, out var parsedName, out var parsedValue))
                 {
+                    if(parsedName == null || string.IsNullOrEmpty(parsedName.Value.Value)
```
Comment (Tratcher) [inline src/Http/Shared/CookieHeaderParserShared.cs:31]:
```suggestion
                    if (parsedName == null || string.IsNullOrEmpty(parsedName.Value.Value)
```

File: src/Http/Shared/CookieHeaderParserShared.cs [hunk 3]
```diff
@@ -107,6 +114,13 @@ private static int GetNextNonEmptyOrWhitespaceIndex(StringSegment input, int sta
         // If we have a separator, skip the separator and all following whitespaces. If we support
         // empty values, continue until the current character is neither a separator nor a whitespace.
         separatorFound = true;
+
+        //We accept single separator, but separator at the end of a value is ignored.
+        if (input.Length > 1 && current + 1 == input.Length)
+        {
+            return current;
+        }
+
```
Comment (Tratcher) [inline src/Http/Shared/CookieHeaderParserShared.cs:123]: The tests now pass even if I remove this new segment. The other fix above is sufficient.
```suggestion
```

Comment (korteksz) [reply src/Http/Shared/CookieHeaderParserShared.cs]: Ah yes, I had it already locally but forget to commit then. Now the suggestions are applied, let's see the pipelines.

File: src/Http/Shared/CookieHeaderParserShared.cs [hunk 4]
```diff
@@ -28,8 +28,15 @@ public static bool TryParseValues(StringValues values, IDictionary<string, strin
             {
                 if (TryParseValue(value, ref index, supportsMultipleValues, out var parsedName, out var parsedValue))
                 {
+                    if (parsedName == null || string.IsNullOrEmpty(parsedName.Value.Value)
+                        || parsedValue == null || string.IsNullOrEmpty(parsedValue.Value.Value))
```
Comment (Tratcher) [inline src/Http/Shared/CookieHeaderParserShared.cs:32]: Sorry, just realized this was allocating strings twice.
```suggestion
                    if (parsedName == null || StringSegment.IsNullOrEmpty(parsedName.Value)
                        || parsedValue == null || StringSegment.IsNullOrEmpty(parsedValue.Value))
```

**Review by Tratcher — APPROVED** [review]

File: src/Http/Shared/CookieHeaderParserShared.cs [hunk 5]
```diff
@@ -28,8 +28,15 @@ public static bool TryParseValues(StringValues values, IDictionary<string, strin
...
+                    if (parsedName == null || StringSegment.IsNullOrEmpty(parsedName.Value)
+                        || parsedValue == null || StringSegment.IsNullOrEmpty(parsedValue.Value))
+                    {
+                        // Successfully parsed, but no values.
+                        continue;
+                    }
+
                     // The entry may not contain an actual value, like " , "
-                    store[parsedName.Value.Value!] = Uri.UnescapeDataString(parsedValue.Value.Value!);
+                    store[parsedName.Value.Value] = Uri.UnescapeDataString(parsedValue.Value.Value);
```
Comment (Tratcher) [inline src/Http/Shared/CookieHeaderParserShared.cs:39]:
```suggestion
                    store[parsedName.Value.Value!] = Uri.UnescapeDataString(parsedValue.Value.Value!);
```

**Comment by cnblogs-dudu** [comment]
This PR disappeared in the release .NET 7.0.1, please see [CookieHeaderParserShared.cs#L29](https://github.com/dotnet/aspnetcore/blob/v7.0.1/src/Http/Shared/CookieHeaderParserShared.cs#L29)

**Comment by ogxd** [comment]
I believe this pull request introduced a small breaking change. I did not test on this repo source directly but from my own code. Here is a repro:
```csharp
[Fact]
public void Repro()
{
  var type = Type.GetType("Microsoft.Net.Http.Headers.CookieHeaderParserShared, Microsoft.Net.Http.Headers");
  var method = type.GetMethod("TryParseValues", BindingFlags.Public | BindingFlags.Static);
  var strValue = new StringValues("cookie1=123;cookie2="); // <- last cookie has no value
  var dic = new Dictionary<string, string>();
  method.Invoke(null, [strValue, dic, true]);
  Assert.True(dic.Count == 2);
}
```
This code passes in .NET 6 and .NET 7, but does not on .NET 8.
Since this PR (I believe), the last cookie with an empty value is ignored, and the dictionary count is now 1.

I am not sure what the behaviour should be to match the W3C as close as possible but:
- It is a breaking change, likely collateral to this PR
- If we ever want to create a cookie with an empty value, it is not possible anymore (but should this be a thing in the first place? I don't know).

**Comment by Tratcher** [comment]
Comments on closed issues are not tracked, please open a new issue with the details for your scenario.

**Pull request merged**